Makefile

```
@@ -1,16 +1,84 @@
+.PHONY: all build test clean install proto
+all: build
+
+# HELP sourced from https://gist.github.com/prwhite/8168133
+
+# Add help text after each target name starting with '\#\#'
+# A category can be added with @category
+
+HELP_FUNC = \
+    %help; \
+    while(<>) { \
+        if(/^([a-z0-9_-]+):.*\#\#(?:@(\w+))?\s(.*)$$/) { \
+            push(@{$$help{$$2}}, [$$1, $$3]); \
+        } \
+    }; \
+    print "usage: make [target]\n\n"; \
+    for ( sort keys %help ) { \
+        print "$$_:\n"; \
+        printf("  %-30s %s\n", $$_->[0], $$_->[1]) for @{$$help{$$_}}; \
+        print "\n"; \
+    }
+
+help:           ##@help show this help
+	@perl -e '$(HELP_FUNC)' $(MAKEFILE_LIST)
+
 NAME="github.com/goto/compass"
 VERSION=$(shell git describe --always --tags 2>/dev/null)
 COVERFILE="/tmp/compass.coverprofile"
 PROTON_COMMIT := "a6b2821e8ddd1127a63d3b376f860990d58931da"
-.PHONY: all build test clean install proto
 
-all: build
+TOOLS_MOD_DIR = ./tools
+TOOLS_DIR = $(abspath ./.tools)
 
-build: ## Build the compass binary
-	go build -ldflags "-X ${NAME}/cli.Version=${VERSION}" 
+$(TOOLS_DIR)/buf: $(TOOLS_MOD_DIR)/go.mod $(TOOLS_MOD_DIR)/go.sum $(TOOLS_MOD_DIR)/tools.go
+	cd $(TOOLS_MOD_DIR) && \
+	go build -o $(TOOLS_DIR)/buf github.com/bufbuild/buf/cmd/buf
+
+$(TOOLS_DIR)/golangci-lint: $(TOOLS_MOD_DIR)/go.mod $(TOOLS_MOD_DIR)/go.sum $(TOOLS_MOD_DIR)/tools.go
+	cd $(TOOLS_MOD_DIR) && \
+	go build -o $(TOOLS_DIR)/golangci-lint github.com/golangci/golangci-lint/cmd/golangci-lint
+
+$(TOOLS_DIR)/mockery: $(TOOLS_MOD_DIR)/go.mod $(TOOLS_MOD_DIR)/go.sum $(TOOLS_MOD_DIR)/tools.go
+	cd $(TOOLS_MOD_DIR) && \
+	go build -o $(TOOLS_DIR)/mockery github.com/vektra/mockery/v2
+
+$(TOOLS_DIR)/gofumpt: $(TOOLS_MOD_DIR)/go.mod $(TOOLS_MOD_DIR)/go.sum $(TOOLS_MOD_DIR)/tools.go
+	cd $(TOOLS_MOD_DIR) && \
+	go build -o $(TOOLS_DIR)/gofumpt mvdan.cc/gofumpt
+
+$(TOOLS_DIR)/gci: $(TOOLS_MOD_DIR)/go.mod $(TOOLS_MOD_DIR)/go.sum $(TOOLS_MOD_DIR)/tools.go
+	cd $(TOOLS_MOD_DIR) && \
+	go build -o $(TOOLS_DIR)/gci github.com/daixiang0/gci
```

this looks messy, let's refactor this
also add an `install` to rule them all, 

```suggestion
define build_tool
$(TOOLS_DIR)/$(1): $(TOOLS_MOD_DIR)/go.mod $(TOOLS_MOD_DIR)/go.sum $(TOOLS_MOD_DIR)/tools.go
	cd $(TOOLS_MOD_DIR) && \
	go build -o $(TOOLS_DIR)/$(1) $(2)
endef

$(eval $(call build_tool,buf,github.com/bufbuild/buf/cmd/buf))
$(eval $(call build_tool,golangci-lint,github.com/golangci/golangci-lint/cmd/golangci-lint))
$(eval $(call build_tool,mockery,github.com/vektra/mockery/v2))
$(eval $(call build_tool,gofumpt,mvdan.cc/gofumpt))
$(eval $(call build_tool,gci,github.com/daixiang0/gci))

install: $(TOOLS_DIR)/buf $(TOOLS_DIR)/golangci-lint $(TOOLS_DIR)/mockery $(TOOLS_DIR)/gofumpt $(TOOLS_DIR)/gci
	@echo "All tools installed successfully"
```

---

Having individual targets allows us to install each command for required Makefile target. Like `buf` for `make proto`, `golangci-lint` for `make lint` etc. Would want to preserve this property.

---

Do we really need `install` target in this case? Each tool gets installed when required.

---

yes, let's keep individual target,
install really helpful for initial local setup and peace of mind,
so next time we want to run any command like lint or fmt, all the tools are available without worrying about internet